### PR TITLE
DEP: deprecate scalar conversions for arrays with ndim > 0

### LINKED ISSUE #10404: deprecate scalar conversions for rank>0 arrays
Numpy allows for conversion of arrays into scalars if they are  size-1, i.e.,
```python
float(numpy.array(1.0))  # 1.0
float(numpy.array([1.0]))  # 1.0
float(numpy.array([[[[1.0]]]]))  # 1.0

# TypeError: only size-1 arrays can be converted to Python scalars
float(numpy.array([1.0, 1.0]))
```
I'm fine with rank-0 conversions, but I found that discrimination of conversion based on array size can easily lead to inconsistencies downstream. See, for example, https://github.com/sympy/sympy/issues/13924 where suddenly you have a different behavior of multiplication depending on the length of the input array.

Moreover, this feature is actually redundant for rank>0 arrays since the value can simply be retrieved from the array via `x[0]`.

When grepping for `"converted to Python scalar"` in the numpy sources, one finds that `numpy/lib/user_array.py` has
```
"only rank-0 arrays can be converted to Python scalars."
```
which seems to be the more consistent solution.

I would hence like to suggest to deprecate scalar conversion for rank>0 size-1 arrays.

### PR DESCRIPTION
**Pull request by nschloe**
This PR reflects some of the progress achieved in issue #10404 and is used to asses the impact of the changes. 

With the changes in this PR, `float(numpy.array([1.0])` now gives a warning; likewise some other things:
```python
import numpy

a = numpy.random.rand(10, 1)
a[0] = numpy.array([1.0])       # okay
a[0] = numpy.array(1.0)         # okay
a[0] = 1.0                      # okay

b = numpy.random.rand(10)
b[0] = numpy.array([1.0])       # ValueError: setting an array element with a sequence.
b[0, ...] = numpy.array([1.0])  # okay
b[0] = numpy.array(1.0)         # okay
b[0] = 1.0                      # okay
```
This aligns the behavior of numpy arrays with that of lists:
```python
float([3.14]) 
```
```
TypeError: float() argument must be a string or a number, not 'list'
```
```python
import numpy as np

a = np.random.rand(5)
a[0] = [3.14]
```
```
ValueError: setting an array element with a sequence.
```
Fixes #10404.

### CODE REVIEW

File: numpy/core/function_base.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
     y += start
 
     if endpoint and num > 1:
-        y[-1] = stop
+        y[-1,...] = stop
```
Comment (eric-wieser) [inline numpy/core/function_base.py]: Why is this needed? Seems the problem is in the callers.

Comment (nschloe) [reply numpy/core/function_base.py]: If `stop` is a scalar, this works fine without `...`. However, linspace also accepts `Arrayish` inputs for `start` and `stop`. See https://github.com/numpy/numpy/issues/3504.

**Comment by eric-wieser** [comment]
Would be interesting to see how extensive the changes need to be to make CI pass, and whether this impacts the scipy or pandas test suites.

**Comment by eric-wieser** [comment]
Looks like randomgen added more tests that rely on this behavior, should be straightforward to just remove them

**Comment by nschloe** [comment]
I'm on it.

File: numpy/random/_bounded_integers.pyx.in [hunk 2]
```diff
@@ -314,10 +314,15 @@ cdef object _rand_{{nptype}}(object low, object high, object size,
     high_arr = <np.ndarray>np.array(high, copy=False)
     low_ndim = np.PyArray_NDIM(low_arr)
     high_ndim = np.PyArray_NDIM(high_arr)
+    # nschloe, 2019-11-29:
+    # This conditional allows x and [x]. Why stop there? [[[[[x]]]]] also wants to be a
+    # low/high. It'd be more consistent if we just allowed int input for low, high in
+    # the first place, not array_like.
```
Comment (eric-wieser) [inline numpy/random/_bounded_integers.pyx.in]: I'd be inclined to just change the test here, rather than trying to allow an array

File: numpy/testing/_private/utils.py [hunk 3]
```diff
@@ -660,7 +660,6 @@ def assert_approx_equal(actual,desired,significant=7,err_msg='',verbose=True):
     __tracebackhide__ = True  # Hide traceback for py.test
     import numpy as np
 
-    (actual, desired) = map(float, (actual, desired))
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Again, I think this is actually correct, and it's the caller that is wrong

**Comment by nschloe** [comment]
So I've been working on scipy a bit (https://github.com/scipy/scipy/pull/11147) and found that in parts it was quite lax when it came to float conversion of arrays. I'll fix all the inconsistencies there and in panda (if any).

All of what I discovered are actual bugs which is what makes me like this PR here even more. It _will_ break (buggy/inefficient) code though, so there'll perhaps have to be a deprecation process. 

What do you think?

**Comment by eric-wieser** [comment]
To me there's no question that if we do this, it will need to be after a deprecation period. I'm definitely in favor of tracking down bugs in downstream projects with this change, whether or not we decide to merge this. Make sure to link those fixes to here so we can get a feel for the impact.

File: numpy/core/src/multiarray/methods.c [hunk 4]
```diff
@@ -2577,8 +2577,9 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     PyArrayObject *arr;
     PyArray_Descr *dtype;
     PyObject *c;
-    if (PyArray_SIZE(self) != 1) {
-        PyErr_SetString(PyExc_TypeError, "only length-1 arrays can "\
+
+    if (PyArray_NDIM(self) != 0) {
+        PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "\
                         "be converted to Python scalars");
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c]: Responding to @seberg's question in the corresponding issue - this, and the other instance, should be:
```
if (PyArray_NDIM(self) != 0) {
    if (PyArray_SIZE(self) == 1) {
        if (DEPRECATE("Conversion of an array with one element to a scalar is deprecated, and will error in future") < 0) {
            return NULL;
        }
    }
    else {
        PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "
                        "be converted to Python scalars");
        return NULL;
    }
}
```

Comment (seberg) [reply numpy/core/src/multiarray/methods.c]: Yeah, when I read the stuff, I thought of `PyArray_Result` first, thats why I thought it would be super complicated. This is fine, it is mostly about how many downstream pain credits we can/should spend on it.

File: numpy/core/src/multiarray/methods.c [hunk 5]
```diff
@@ -2582,8 +2582,8 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
         if (PyArray_SIZE(self) == 1) {
             if (DEPRECATE(
                 "Conversion of an array with one element to a scalar \
-                is deprecated, and will error in future. \
-                Ensure you extract a single element from your array before performing this operation"
+is deprecated, and will error in future. \
+Ensure you extract a single element from your array before performing this operation"
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c]: You're looking for
```C
            if (DEPRECATE(
                "Conversion of an array with one element to a scalar "
                "is deprecated, and will error in future. "
                "Ensure you extract a single element from your array before performing this operation."
            ) < 0) {
```

File: numpy/random/_bounded_integers.pyx.in (hunk 2 above)
Comment (nschloe) [reply numpy/random/_bounded_integers.pyx.in]: The documentation says: `low : int or array-like` (likewise `high`). Should I change that?

File: numpy/core/src/multiarray/methods.c (hunk 4 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/methods.c]: Probably worthwhile extracting a helper function so this code isn't duplicated.

File: numpy/random/_bounded_integers.pyx.in (hunk 2 above)
Comment (eric-wieser) [reply numpy/random/_bounded_integers.pyx.in]: Yes

Comment (eric-wieser) [reply numpy/random/_bounded_integers.pyx.in]: Perhaps remove this weirdness in a separate pr?

Comment (eric-wieser) [reply numpy/random/_bounded_integers.pyx.in]: Ah, hang on - this is just an optimisation. There's a full broadcast path below. I'd just remove this path for all but the 0d case

Comment (nschloe) [reply numpy/random/_bounded_integers.pyx.in]: Right now, it's possible to do `randint([0, 1], [4, 7])` which gives two ints, one between 0 and 4 and another between 1 and 7. This has been introduced sometime after 1.16. I've personally never used this but I'm not sure if this is a common use case. We could allow in arrays of arbitrary shape and have the output of shape `low_shape + size` (`low_shape == high_shape` of course).

Comment (nschloe) [reply numpy/random/_bounded_integers.pyx.in]: > I'd just remove this path for all but the 0d case

Indeed.

File: numpy/core/src/multiarray/common.c [hunk 6]
```diff
@@ -945,5 +945,24 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
     }
 }
 
-
-
+bool
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.c]:
```suggestion
NPY_NO_EXPORT bool
```

File: numpy/core/src/multiarray/common.c [hunk 7]
```diff
@@ -945,5 +945,24 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
...
+has_rank_0(PyArrayObject *v)
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.c]: Probably better to call this
```suggestion
check_has_rank_0(PyArrayObject *v)
```
and give it a `int` return of either `0` or `-1`, to make it clear that it sets an error

File: numpy/core/tests/test_multiarray.py [hunk 8]
```diff
@@ -7578,9 +7582,10 @@ def test_to_int_scalar(self):
         # gh-9972 means that these aren't always the same
         int_funcs = (int, lambda x: x.__int__())
         for int_func in int_funcs:
-            assert_equal(int_func(np.array([1])), 1)
-            assert_equal(int_func(np.array([0])), 0)
-            assert_equal(int_func(np.array([[42]])), 42)
+            assert_equal(int_func(np.array(1)), 1)
+            assert_equal(int_func(np.array(0)), 0)
+            assert_raises(DeprecationWarning, int_func, np.array([1]))
+            assert_raises(DeprecationWarning, int_func, np.array([[42]]))
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: This should be:
```
with pytest.assert_warns(DeprecationWarning):
    assert_equal(int_func(np.array([[42]])), 42)
```
or similar

File: numpy/core/tests/test_multiarray.py [hunk 9]
```diff
@@ -3299,10 +3299,8 @@ def test__complex__(self):
             msg = 'dtype: {0}'.format(dt)
             ap = complex(a)
             assert_equal(ap, a, msg)
-            bp = complex(b)
-            assert_equal(bp, b, msg)
-            cp = complex(c)
-            assert_equal(cp, c, msg)
+            assert_warns(DeprecationWarning, complex, b)
+            assert_warns(DeprecationWarning, complex, c)
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: You should be able to simultaneously assert that this warns and check its value here

File: numpy/core/tests/test_multiarray.py [hunk 10]
```diff
@@ -7578,9 +7582,11 @@ def test_to_int_scalar(self):
         # gh-9972 means that these aren't always the same
         int_funcs = (int, lambda x: x.__int__())
         for int_func in int_funcs:
-            assert_equal(int_func(np.array([1])), 1)
-            assert_equal(int_func(np.array([0])), 0)
-            assert_equal(int_func(np.array([[42]])), 42)
+            assert_equal(int_func(np.array(1)), 1)
+            assert_equal(int_func(np.array(0)), 0)
+            with assert_warns(DeprecationWarning):
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: You need a with block for each function call, so that you can prove that _both_ warn

**Comment by nschloe** [comment]
pandas builds and tests cleanly against this PR.

File: numpy/core/src/multiarray/methods.c [hunk 11]
```diff
@@ -2577,10 +2577,9 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     PyArrayObject *arr;
     PyArray_Descr *dtype;
     PyObject *c;
-    if (PyArray_SIZE(self) != 1) {
-        PyErr_SetString(PyExc_TypeError, "only length-1 arrays can "\
-                        "be converted to Python scalars");
-        return NULL;
+
+    if (check_has_rank_0(self) < 0) {
+      return NULL;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/methods.c]:
```suggestion
        return NULL;
```

File: numpy/core/tests/test_multiarray.py [hunk 12]
```diff
@@ -3323,9 +3328,15 @@ def test__complex__should_not_work(self):
...
 
-        e = np.array(['1+1j'], 'U')
+        e = np.array('1+1j', 'U')
         assert_raises(TypeError, complex, e)
 
+        # This gives both a TypeError (because of the 'U') and a DeprecationWarning
+        # (because of the []). Not sure how to deal with it.
+        # e = np.array(['1+1j'], 'U')
+        # assert_raises(TypeError, complex, e)
+        # assert_warns(DeprecationWarning, complex, e)
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: I think you can use `assert_warns` wrapping the original `assert_raises`

File: numpy/random/_bounded_integers.pyx.in [hunk 13]
```diff
@@ -314,8 +314,7 @@ cdef object _rand_{{nptype}}(object low, object high, object size,
     high_arr = <np.ndarray>np.array(high, copy=False)
     low_ndim = np.PyArray_NDIM(low_arr)
     high_ndim = np.PyArray_NDIM(high_arr)
-    if ((low_ndim == 0 or (low_ndim == 1 and low_arr.size == 1 and size is not None)) and
-            (high_ndim == 0 or (high_ndim == 1 and high_arr.size == 1 and size is not None))):
+    if (low_ndim == 0 and high_ndim == 0):
```
Comment (eric-wieser) [inline numpy/random/_bounded_integers.pyx.in]: nit:
```suggestion
    if low_ndim == 0 and high_ndim == 0:
```

Comment (eric-wieser) [reply numpy/random/_bounded_integers.pyx.in]: Want to spin out a PR with this fixup in to be merged immediately?

File: numpy/testing/tests/test_utils.py [hunk 14]
```diff
@@ -609,8 +609,8 @@ def setup(self):
         self._assert_func = assert_approx_equal
 
     def test_simple_arrays(self):
```
Comment (eric-wieser) [inline numpy/testing/tests/test_utils.py]: Since this was never intended to be used on 1d arrays:
```suggestion
     def test_simple_0d_arrays(self):
```

File: numpy/testing/tests/test_utils.py [hunk 15]
```diff
@@ -609,8 +609,8 @@ def setup(self):
...
-        x = np.array([1234.22])
-        y = np.array([1234.23])
+        x = np.array(1234.22)
+        y = np.array(1234.23)
```
Comment (eric-wieser) [inline numpy/testing/tests/test_utils.py]: Want to spin out a PR with this fixup in to be merged immediately?

File: numpy/core/tests/test_multiarray.py [hunk 16]
```diff
@@ -7578,9 +7589,12 @@ def test_to_int_scalar(self):
         # gh-9972 means that these aren't always the same
         int_funcs = (int, lambda x: x.__int__())
         for int_func in int_funcs:
-            assert_equal(int_func(np.array([1])), 1)
-            assert_equal(int_func(np.array([0])), 0)
-            assert_equal(int_func(np.array([[42]])), 42)
+            assert_equal(int_func(np.array(1)), 1)
+            assert_equal(int_func(np.array(0)), 0)
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: These two lines could go into the spin-off PR too

File: numpy/core/src/multiarray/common.c [hunk 17]
```diff
@@ -945,5 +945,24 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
...
+            }
+        }
+        else {
+            PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "
+                            "be converted to Python scalars");
+            return -1;
+        }
+    }
+    return 0;
+}
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.c]: super nit: could eliminate some nesting with
```C
check_has_rank_0(PyArrayObject *v)
{
    if (PyArray_NDIM(v) == 0) {
        return 0;
    }
    /* Remove this if when the deprecation expires */
    if (PyArray_SIZE(v) == 1) {
        /* Numpy 1.19.0, 2019-12-18 */
        if (DEPRECATE(
            "Conversion of an array with one element to a scalar "
            "is deprecated, and will error in future. "
            "Ensure you extract a single element from your array before performing this operation."
        ) < 0) {
            return -1;
        }
        return 0;
    }
    PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "
                        "be converted to Python scalars");
    return -1;
}
```

**Comment by eric-wieser** [comment]
One failure remains:

https://dev.azure.com/numpy/numpy/_build/results?buildId=6980&view=ms.vss-test-web.build-test-results-tab&runId=1082926&resultId=101493&paneView=debug

**Comment by nschloe** [comment]
Added to #15126.

**Comment by nschloe** [comment]
Right now looking at pandas. (I reported earlier that it builds cleanly, but that's not the case -- You have to run the tests with `-Werror`.)

**Comment by eric-wieser** [comment]
It's reassuring how few changes were required in #15126 and our test suite.

**Comment by nschloe** [comment]
matplotlib requires one change for the tests to pass without warning. https://github.com/matplotlib/matplotlib/pull/16003

**Comment by seberg** [comment]
It sounds like we probably won't be pushing this forward right now (unfortunately). The issue is that it seems like probably have to spend a lot of "breaks user code" chips on it, and not everyone is even quite convinced that it is incorrect, as opposed to just slightly weird semantics. Sorry about this, but it sounds like the chance is that we will be closing this soon for now :/.

My opinion: Once all the major packages have been fixed for a while, and with a few examples of how this creates actual bugs, we may be able to reconsider (I would actually like that). But I do not have examples of bugs and write it up for the mailing list. If we can convince that deprecating this (very very slowly) fixes some bugs and is likely not to annoy too many users out there, there is a case that can be made.

**Comment by charris** [comment]
@seberg This should be closed?

**Comment by seberg** [comment]
I am not sure. I still think it is desirable, but whether or not we should go ahead depends a bit on how the major packages would break with it. @nschloe do you remember the status?

**Comment by nschloe** [comment]
> I am not sure. I still think it is desirable, but whether or not we should go ahead depends a bit on how the major packages would break with it. nschloe do you remember the status?

At the the time, I probed some of the biggest NumPy-dependent Python packages (scipy, sympy, mpl, astropy I think) and submitted a bunch of PRs. Not sure of the status there, but I can check again.

I also think this is still desirable, so perhaps this can be left open as an "epic" of sorts.

**Comment by eric-wieser** [comment]
I remember large chunks of scipy that were sloppy about this, but I think a motivated user could fix them one by one. So I'd agree with leaving this open, in case anyone with such motivation appears.

**Comment by nschloe** [comment]
- [ ] matplotlib https://github.com/matplotlib/matplotlib/pull/19952

**Comment by nschloe** [comment]
Whenever I get back to scipy, open an issue or even worse, a PR, it makes me appreciate the numpy community so much more. So thanks guys for sticking around. :+1:

**Comment by nschloe** [comment]
> I remember large chunks of scipy that were sloppy about this, but I think a motivated user could fix them one by one. 

I'm happy to report that said motivated individual has plowed through the scipy issues and fixed all of them in 3 PRs. (See above.) Most of the issues were due to a sloppy handling of goal functionals not returning actual scalars. All the changes have improved the scipy code.

I had already fixed matplotlib and pandas a while ago; the changes there where miniscule.

I would consider this PR ready for merge and would be happy about a re-review.

**Comment by mattip** [comment]
This still needs a release note

File: numpy/core/src/multiarray/common.c [hunk 18]
```diff
@@ -476,3 +476,25 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
...
+            "Ensure you extract a single element from your array before performing this operation."
+        ) < 0) {
+            return -1;
+        }
+        return 0;
+    }
+    PyErr_SetString(PyExc_TypeError, "only rank-0 arrays can "
+                        "be converted to Python scalars");
+    return -1;
+}
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.c:500]:
```suggestion
}

```

File: numpy/core/src/multiarray/common.h [hunk 19]
```diff
@@ -345,3 +345,5 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
 
 #endif
 
+NPY_NO_EXPORT int
+check_has_rank_0(PyArrayObject *v);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/common.h:349]:
```suggestion
check_has_rank_0(PyArrayObject *v);

```

**Comment by nschloe** [comment]
@mattip Release note added.

File: numpy/core/src/multiarray/common.c [hunk 20]
```diff
@@ -476,3 +476,25 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
...
 
+NPY_NO_EXPORT int
+check_has_rank_0(PyArrayObject *v)
+{
+    if (PyArray_NDIM(v) == 0) {
+        return 0;
+    }
+    /* Remove this if when the deprecation expires */
+    if (PyArray_SIZE(v) == 1) {
+        /* Numpy 1.19.0, 2019-12-18 */
```
Comment (h-vetinari) [inline numpy/core/src/multiarray/common.c:487]: This line likely needs updating too...

Comment (nschloe) [reply numpy/core/src/multiarray/common.c]: Thanks, fixed now.

File: numpy/core/src/multiarray/common.c [hunk 21]
```diff
@@ -484,7 +484,7 @@ check_has_rank_0(PyArrayObject *v)
     }
     /* Remove this if when the deprecation expires */
     if (PyArray_SIZE(v) == 1) {
-        /* Numpy 1.19.0, 2019-12-18 */
+        /* Numpy 1.21.3, 2021-08-17 */
```
Comment (h-vetinari) [inline numpy/core/src/multiarray/common.c:487]: Not to be too picky, but I presume this is not going to be backported, so should probably say `1.22.0`.

Comment (nschloe) [reply numpy/core/src/multiarray/common.c:487]: Ah yes. Should the deprecation note go in `doc/source/release/1.22.0-notes.rst`, too? (That file is still more or less empty.)

Comment (seberg) [reply numpy/core/src/multiarray/common.c:487]: It should go as a deprecation into `doc/release/upcoming_changes`, check the `README` in that file.  The release note is incorrect as is.

**Review by seberg — COMMENTED** [review]
It would be good to have a test in `core/tests/test_deprecations.py`.  It is not super important to be in that file, but it keeps deprecations in one place and also ensures that we test the error path.

(I don't see code coverage complaining, although I am not sure what would be covering the error paths).

File: numpy/core/src/multiarray/common.c [hunk 22]
```diff
@@ -476,3 +476,25 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
...
+        return 0;
+    }
+    /* Remove this if when the deprecation expires */
+    if (PyArray_SIZE(v) == 1) {
+        /* Numpy 1.19.0, 2019-12-18 */
+        if (DEPRECATE(
+            "Conversion of an array with ndim > 0 to a scalar "
+            "is deprecated, and will error in future. "
+            "Ensure you extract a single element from your array before performing this operation."
+        ) < 0) {
```
[the remark this answers is not in the document]
Comment (nschloe) [reply numpy/core/src/multiarray/common.c]: Fixed.

**Comment by nschloe** [comment]
> It would be good to have a test in core/tests/test_deprecations.py. It is not super important to be in that file, but it keeps deprecations in one place and also ensures that we test the error path.
> (I don't see code coverage complaining, although I am not sure what would be covering the error paths).

I added some explicit deprecation tests now. The previous coverage came from some `assert_warns(DeprecationWarning)` in `numpy/core/tests/test_multiarray.py` (see diff).

**Comment by nschloe** [comment]
If there's anything I can do to advance this, let me know.

**Comment by seberg** [comment]
While I personally like this, it might also be a bit of my bias towards consistency vs. some (rare/non rare?) convenience.  Could you hit the mailing list outlining the proposal and what it improves and breaks?

I tend towards trying it and roll-back if there is push-back, but I am not sure it is important enough to fight hard for it.

**Comment by nschloe** [comment]
@asmeurer came up with an interesting example. It goes something like
```python
import numpy as np

a = np.random.rand(2, 3)
a[0] = [1, 2, 3]                # ok
a[0] = np.array([1, 2, 3])      # ok
# a[0] = [[[[1,2,3]]]]          # ValueError: setting an array element with a sequence.
a[0] = np.array([[[[1,2,3]]]])  # ok?!
```
The last line does something like a squeeze before setting the array, and is not intercepted by the changes in this PR. It can be thought of an exension of the
```python
a = np.random(3)
a[0] = np.array([1])
```
that we'd like to deprecate. I'm not sure how common assignments like `a[0] = np.array([[[[1,2,3]]]])` are, but I'm leaning towards deprecating it too for the sake of consistency with lists and consistency with the deprecation in this PR. Not sure if it should be done as part of this PR. What do y'all think?

**Comment by nschloe** [comment]
> I was aware of it, and always considered this part of the deprecation here. I am not sure why you see it as different from your example of arr1d[0] = [1.]?

The only difference is that
```python
import numpy as np

a = np.random.rand(2, 3)
a[0] = np.array([[[[1,2,3]]]])
```
does _not_ yield a warning even after this PR. Question: Should it be added in another PR, or in this?

**Comment by seberg** [comment]
Oh, I thought this PR would affect that as well!  Maybe the reason why it was so painless;  so there may be a point in not including it just to make it a smaller, less painful, step...

**Comment by nschloe** [comment]
Nah, just scalars for now. I think that's already an improvement over the current behavior. I think it'd make sense to have the multidim case follow in a second PR.

**Comment by Micky774** [comment]
@nschloe Are you still interested in continuing this PR? If so, I have a couple modifications that reconcile changes w/ upstream main and fix the failing tests that I can open a PR to your branch with. If you're not interested in / able to continue this PR, I wouldn't mind picking it up :)

**Comment by nschloe** [comment]
I sure am! Not sure what exactly needs to be done though. Perhaps another run on SciPy, Pandas, etc. to see if anything has crept up in the meantime.

**Comment by Micky774** [comment]
> I sure am! Not sure what exactly needs to be done though. Perhaps another run on SciPy, Pandas, etc. to see if anything has crept up in the meantime.

Hey there! I hadn't heard from you in a while so I went ahead and opened a follow-up PR #21846 to ensure that the great work you did would make it through to merge. I preserved your commits so you retain credit in the eventual squash-merge authorship. Not sure how the maintainers want to reconcile this, but feel free to hop over there and offer some feedback on the modifications I made as well :)

**Comment by nschloe** [comment]
> Hey there! I hadn't heard from you in a while so I went ahead and opened a follow-up PR

You opened that about 12 hours after you asked here.

**Comment by Micky774** [comment]
@nschloe I'm happy to close my PR. The reason I opened my PR so quickly after asking was because I saw that there was no activity on this PR for ~7 months already, and because I wouldn't mind closing it if you were interested in continuing this one. 

I just want to make it clear that this wasn't in interest of claiming credit for your work -- in the main body of my PR I laid out explicitly that the significant contributions were yours and that mine were minor reconciliations. Sorry if this caused you to worry that the work you did was going to be undermined. Again, it really is good work and I apologize that my actions caused an unnecessary stir.

**Comment by mattip** [comment]
There are some failing tests.

File: numpy/core/src/multiarray/methods.c [hunk 23]
```diff
@@ -2761,9 +2761,7 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     PyArray_Descr *dtype;
     PyObject *c;
 
-    if (PyArray_SIZE(self) != 1) {
-        PyErr_SetString(PyExc_TypeError,
-                "only length-1 arrays can be converted to Python scalars");
+    if (check_has_rank_0(self) < 0) {
```
Comment (mattip) [inline numpy/core/src/multiarray/methods.c:2764]: This needs to be declared in a header. I wonder how it passes compilation?

**Comment by nschloe** [comment]
So this
```python
arr = np.ones((1, 1))
np.array([0, arr], dtype=np.float64)
```
Results in _two_ deprecation warnings after this PR. One is old,
```
<ipython-input-3-4e22c1d93d88>:1: DeprecationWarning: setting an array element with a sequence. This was supported in some cases where the elements are arrays with a single element. For example `np.array([1, np.array([2])], dtype=int)`. In the future this will raise the same ValueError as `np.array([1, [2]], dtype=int)`.
  np.array([0, arr], dtype=np.float64)
```
and one is new,
```
<ipython-input-3-4e22c1d93d88>:1: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.24.)
  np.array([0, arr], dtype=np.float64)
Out[3]: array([0., 1.])
```
The latter occurs because internally the `arr` is converted to a scalar. It's nice to see the deprecations aligning, but we probably don't want two.

**Comment by seberg** [comment]
Lets not worry about the duplicate.  In fact, I think I may just finalize that deprecation now.

**Comment by nschloe** [comment]
@seberg Cool. Let me know if there's anything I can do to help.

**Comment by seberg** [comment]
We may have to rebase now, since I removed the deprecation in gh-22004 so this should be an error now.

**Comment by nschloe** [comment]
I manually rebased.

**Comment by nschloe** [comment]
@seberg Let me know if there's anything I can do to advance this.

**Comment by nschloe** [comment]
I've updated SciPy again to be warnings-free after this PR (https://github.com/scipy/scipy/pull/17698). Changes are now usually minimal.

I'd be great if this PR could be looked at again. We'll soon be celebrating it's 5th birthday. :)

**Comment by asmeurer** [comment]
Are there plans to also deprecate the related reverse broadcasting behavior which was discussed on the list?

**Comment by mattip** [comment]
I marked this for triage review, it also has a conflict.

**Comment by seberg** [comment]
Planning to put this in later today and give it a shot.  There is always the (correct) worry here that it might be noisy for downstream.  OTOH, this has also repeatedly come up over the time.

We had brought it up very briefly yesterday, but @mattip, if you had a different intention when marking it, just let me know.

What makes me optimistic is that the scipy changes are far larger than the other libraries (all of which seem very simple, the optimizers really were most of it probably).

So to be clear:  I am happy to roll it back if this annoys downstream libraries or anyone has serious concerns (libs like dask or xarray for example).
As for downstream users, I don't mind taking this deprecation slow.

**Comment by mroeschke** [comment]
@seberg in pandas, how should we treat size 1 arrays with PyObject list/tuples that we want to treat as scalars for assignment e.g.

```
-> values[indexer] = casted
(Pdb) values
array([10, 11, 12], dtype=object)
(Pdb) indexer
array([1])
(Pdb) casted
array([(61,)], dtype=object)
(Pdb) n
(Pdb) values
array([10, (61,), 12], dtype=object)  # want to assign a tuple
(Pdb) np.__version__
'1.24.2'

---

(Pdb) values[indexer] = casted[0]  # correct workaround for the deprecation?
(Pdb) values
array([10, 61, 12], dtype=object)  # 61 is assigned instead of (61,)
```

**Comment by asmeurer** [comment]
@mroeschke you can work around that by doing `values[indexer] = casted[0, ...]`. That forces the right-hand side to be a 0-D array instead of a scalar, which prevents NumPy from trying to call asarray on it.

**Comment by asmeurer** [comment]
Even this deprecation notwithstanding it's a good idea to use 0-D arrays instead of scalars everywhere if you're dealing with object dtype arrays with elements that themselves could be interpreted as arrays by `asarray`. 

The issue is that `casted[0]` is literally type `tuple` (i.e., not a NumPy type anymore), so there's inherent ambiguity whether `(61,)` should be interpreted as the tuple object or as the 1-D array `array([61])` in pretty much any context where NumPy could call `asarray` (which is virtually everywhere).

**Comment by neutrinoceros** [comment]
Provided my reading of the patch, and in particular the comment attached to the deprecation code, is correct, the error that will be raised after this deprecation is expired is `TypeError` ?

For context, I'm working on a case where the implicit conversion from size-1 array to scalar happens in a `try: ... except TypeError:` block, and landing in the `except` branch would be acceptable.

**Comment by seberg** [comment]
Yes, it will be a `TypeError`.
